STN-181: Vertical Link Card

kaseybon opened the pull request:
# READY FOR REVIEW

## Summary
This PR adds the vertical link card and styles. The only required field in a vertical link card is the title so various combinations of title, image, and description test have been tested:

<img width="993" alt="Screen Shot 2020-03-13 at 5 11 56 PM" src="https://user-images.githubusercontent.com/2486846/76659862-e2317280-654d-11ea-9ec7-9c106bbfad1f.png">

Design: https://www.figma.com/file/toNZnFLadtpNpG5dPzNGWz/%E2%9A%99%EF%B8%8F-Colorful-Theme---Working?node-id=4321%3A146

Has been design reviewed by @meranicosme 

*Note: There are some responsive issues where things look extra squished at certain breakpoints. This is more of a grid problem and should be resolved in a design polish card*

_More Notes: We've made a config update at the platform level so that vertical linked cards will always return a url for the button variable. This update clashes with how the Stanford Humsci theme works and will cause vertical cards to break for older themes. To ensure that does not happen we have also removed the function in Stanford Humsci that was updating the button variable._

**Browser Tested In:**
- Chrome (Windows & Mac)
- Firefox (Windows & Mac)
- Safari
- Edge
- IE 11 (There is some breakage but it appears to be related to the grid, not the card)

## Need Review By (Date)
3/16

## Urgency
medium

## Steps to Test
1. Run `npm run build:sass` to build CSS assets
1. Run `npm test` to confirm there are no linting errors
1. Test using as a paragraph both in a row and outside of a row
1. Optionally this can be used as a view:
    - Duplicate the vertical card view
    - Set the Pattern to _Vertical Link Card_
    - Map the following fields:
        - Title -> Title
        - Image -> Image
        - Body -> Description
        - Link to Content -> Button
    - Edit the Title field to render in an `<h2>` 
    - Update the Link to Content field and check the `Output the URL as text` box
1. Test to make sure we are not breaking other themes:
    - Switch to the Stanford Humsci theme and ensure that vertical link cards do not break
    - Check that horizontal and vertical cards are unaffected by the change

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)

kaseybon committed:
feat: vertical link card styles

STN-181

silverli commented on docroot/themes/humsci/humsci_basic/patterns/vertical-link-card/vertical-link-card.html.twig:
```diff
@@ -1,28 +1,27 @@
 {% set attributes = attributes.addClass('vertical-link-card') %}
```
should it be `hb-vertical-linked-card`?

silverli commented on docroot/themes/humsci/humsci_basic/patterns/vertical-link-card/vertical-link-card.html.twig:
```diff
@@ -1,28 +1,27 @@
 
   {% if button %}
-    </a>
+    <a href="{{ button|render|striptags|trim }}" aria-label="{{ title|render|striptags|trim }}" class="vertical-link-card__link"></a>
```
hmm I think something is still off here, I am getting the external link text here instead of the url. I would take a look at the horizontal or vertical card implementation to see if there's a difference

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_vertical-link-card.scss:
```diff
@@ -0,0 +1,112 @@
+  @include hb-colorful {
+    color: $hb-color--white;
+
+    // The gradient adds a 20px hight white block at the top of the linked card
```
hight -> height

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_vertical-link-card.scss:
```diff
@@ -0,0 +1,112 @@
+    // The gradient adds a 20px hight white block at the top of the linked card
+    // so that the title looks like it has a negative margin on the top when
+    // there isn't an image.
+    background-image: linear-gradient(to bottom, $hb-color--white, $hb-color--white hb-calculate-rems(20px), darken(hb-colorful-variation(tertiary), 30%) hb-calculate-rems(20px));
```
can you try spacing this more like so (might also help with it looking intimidating:

```
linear-gradient(
   to bottom,
  $hb-color--white,
  $hb-color--white hb-calculate-rems(20px),
  darken(hb-colorful-variation(tertiary),
  30%) hb-calculate-rems(20px)
);
```

kaseybon replied:
I'm pretty sure our scss linter is fine with breaking the gradient onto multiple lines but the CodeClimate linter will errorr.

silverli replied:
no problemo

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_vertical-link-card.scss:
```diff
@@ -0,0 +1,112 @@
+
+    img {
+      width: 100%;
+      transform: scale(1);
```
do we have a mixin for this image zoom style we use in a couple places? might be nice to add for consistency

kaseybon replied:
I looked into adding mixins but unfortunatly there are too many moving pieces and they all have different parent classes and hover triggers. However, the `scale()` on hover can be moved into a variable!

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_vertical-link-card.scss:
```diff
@@ -0,0 +1,112 @@
+  &__link {
+    // We need the entire vertical-link-card to be a link but wrapping the entire
+    // element or wrapping individual elements in anchor tags is not very accessible.
+    // Instead we'll add a link at the bottom of the card (with an aria-label)
```
:)

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_functions.animation.scss:
```diff
@@ -0,0 +1,3 @@
+@function hb-transition($property) {
```
sweet!

silverli reviewed (changes requested):
This is looking really great. A couple comments in the code.

There is also one thing about using a row with postcard paragraph type: it looks like we need a little more margin between the cards:

<img width="808" alt="Screen Shot 2020-03-16 at 2 12 02 PM" src="https://user-images.githubusercontent.com/12848168/76787914-50657780-6790-11ea-888f-11a4f23ee596.png">
<img width="723" alt="Screen Shot 2020-03-16 at 2 12 16 PM" src="https://user-images.githubusercontent.com/12848168/76787915-50657780-6790-11ea-9e43-9321f0aafd6a.png">
<img width="471" alt="Screen Shot 2020-03-16 at 2 14 15 PM" src="https://user-images.githubusercontent.com/12848168/76787992-7559ea80-6790-11ea-8554-f88528879ae3.png">

kaseybon committed:
refactor: update the vertical-linked-card postcard to export the url only for the button variable

- the su-humsci-theme will break with this change since they use a function to grab only the url from buttons so this commit also removes that code

STN-181

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_functions.animation.scss:
```diff
@@ -0,0 +1,3 @@
+@function hb-transition($property) {
+  @return $property $hb-animation-duration $hb-animation-timing-function;
```
nice!! consistency all around

silverli reviewed (approved):
Looks much better in a row now!!

bryanbraun commented:
These changes impact the existing theme, but I think we've covered our bases on ensuring backwards compatibility. Here's a summary of those changes:

There was [a preprocess function](https://github.com/SU-HSDO/suhumsci/pull/523/files#diff-0968c95dc0e351f9496dd84268056b4cL237-L257) that was hand-picking out the URL for use in the vertical linked card template. We could have ported this to the new theme, but we realized that we could just [update the config](https://github.com/SU-HSDO/suhumsci/pull/523/files#diff-163a0e5555fb772b9b06f675868f03d6L89-R89) to pass the URL as text instead, and delete the preprocess function. These changes cancel each other out, and thus shouldn't impact the display of postcards in the existing theme.

(note: removing the `decanter-button` class should also have no impact... it wasn't being used in the first place because the link url was being hand-picked out of the markup in the preprocess function)

The only possible side-effect we thought of was views using the Linked Vertical Card (which would also need to be reconfigured). We talked to the H&S team and they assured us that they weren't using this card in any views, so we should be clear there. 👍
